common/logisticspipes/utils/GuideBookContents.kt

```diff
@@ -8,56 +9,84 @@ import net.minecraft.util.ResourceLocation
-          val title = json.get("title").asString
-          val pages = json.get("page_count").asInt
-          if (pages < 1) {
-            LogisticsPipes.log.error("Guide book for language '$lang' has $pages pages?!")
+class GuideBookContents private constructor(val lang: String, val divisions: List<Division> = mutableListOf(), val title: String) {
+
+    companion object {
+
+        private val rm = Minecraft.getMinecraft().resourceManager
+
+
+        @JvmStatic
+        @JvmOverloads
+        fun load(lang: String = Minecraft.getMinecraft().languageManager.currentLanguage.languageCode): GuideBookContents? {
+            try {
+                val res = rm.getResource(ResourceLocation(LPConstants.LP_MOD_ID, "book/$lang/book.json"))
+                res.use {
+                    val json = JsonParser().parse(InputStreamReader(res.inputStream)).asJsonObject
+                    val title = json.get("title").asString
+                    val divisions = json.get("divisions").asJsonArray.withIndex().map { (sIndex, divs) -> Division(lang, divs.asJsonObject, sIndex) }
+                    return GuideBookContents(lang, divisions, title)
+                }
+            } catch (e: IOException) {
+                if (lang != "en_us") {
+                    LogisticsPipes.log.error("Couldn't find book data for language '$lang'! Falling back to en_us…")
+                    return load("en_us")
+                } else {
+                    LogisticsPipes.log.error("Couldn't find book data for en_us! Something is horribly wrong…")
+                }
+            } catch (e: UnsupportedOperationException) {
+                LogisticsPipes.log.error("Invalid book.json for language '$lang'!", e)
+            }
             return null
-          }
-          return GuideBookContents(lang, pages, title)
         }
-      } catch (e: IOException) {
-        if (lang != "en_us") {
-          LogisticsPipes.log.error("Couldn't find book data for language '$lang'! Falling back to en_us…")
-          return load("en_us")
-        } else {
-          LogisticsPipes.log.error("Couldn't find book data for en_us! Something is horribly wrong…")
+
+
+    }
+
+    fun getDivision(index: Int): Division{
+        return divisions[index]
+    }
+
+    fun getChapter(dindex: Int, index: Int): Chapter?{
+        return getDivision(dindex).getChapter(index)
+    }
+
+    fun getPage(dindex: Int, cindex: Int, index: Int): Page?{
+        return getDivision(dindex).getChapter(cindex)?.getPage(index)
+    }
+
+    class Page(val dindex:Int, val cindex: Int, val index: Int, val text: String)
+
+    class Chapter(private val lang: String, val parentTitle: String, json: JsonObject, val dindex: Int, val cindex: Int) {
+        val title = json["title"].asString
+        val item = json["item"].asString
+        val nPages = json["pages"].asInt
+
+        fun getPage(index: Int): Page? {
+            var par = parentTitle.replace("[^A-z]".toRegex(), "_").toLowerCase()
```

Comment:
Don't get the file path from the title, or the page will need to be renamed whenever the title changes, which is not ideal. Also this prevents consistent file names across localizations.

Reply:
Yes that is already in the set to be done, unfortunately I've been having problems with stuff not working so those have taken priority

Reply:
Added to Glo board